### DISCUSSION TITLE
怎么才能不去推测要链接的文件？

### DISCUSSION THREAD

**Discussion by batu1579** [xmake-io/xmake#5957] [Q&A] [answered]
我目前编写的程序需要用到一个叫 pylon 的库，这个库提供的预编译版本只有 MD 版本的，我目前用下面的脚本添加这个库的所有文件：

```lua
-- Pylon

local pylon_path = "third_party/pylon"
local pylon_lib_path = pylon_path .. "/lib/x64"
local pylon_include_path = pylon_path .. "/include"
local pylon_runtime_path = pylon_path .. "/Runtime/x64"

-- add_linkdirs(pylon_lib_path)

local pylon_lib_files = os.files(path.join(pylon_lib_path, "*.lib"))
for _, lib_file in ipairs(pylon_lib_files) do
    local lib_name = path.basename(lib_file):match("(.*)%.lib$")
    add_links(lib_name)
end
```

这些库文件的名字有一些类似 `GCBase_MD_VC141_v3_1_Basler_pylon.lib` 这样的格式。但是当我切换为 debug 模式时，编译器会提示我：

```
1>EXEC : error : LINK : fatal error LNK1104: 无法打开文件“GCBase_MDd_VC141_v3_1_Basler_pylon.lib”
```

我当中的 `MD` 被修改为了 `MDd` 但是事实上这个文件并不存在，貌似是当我改为 debug 模式后自动推测出来的。这个 pylon 库并没有提供 MDd 参数的预编译文件，我该怎么样才能避免这种情况发生，正常使用 debug 模式开发？

**Comment by star-hengxing** [comment, marked as answer]
msvc 一个功能：https://en.wikipedia.org/wiki/Auto-linking

如果能像 boost 一样，有一个宏 `BOOST_ALL_NO_LIB` 可以关闭更好，否则只能手动去掉头文件里 auto-link 代码。

**Reply by batu1579** [reply]
我在头文件中搜索了一下 `pragma comment` ，找到了提前定义好的两个宏，添加上以后就可以正常编译了！

```lua
    add_defines("PYLON_NO_AUTO_IMPLIB")
    add_defines("GENICAM_NO_AUTO_IMPLIB")
```

实在是太感谢啦！
